crates/goose/src/providers/litellm.rs

```diff
@@ -0,0 +1,364 @@
+        let url = base_url.join(&self.base_path).map_err(|e| {
+            ProviderError::RequestFailed(format!("Failed to construct endpoint URL: {e}"))
+        })?;
+
+        let request = self
+            .client
+            .post(url)
+            .header("Authorization", format!("Bearer {}", self.api_key));
+
+        let request = self.add_headers(request);
+
+        let response = request.json(&payload).send().await?;
+
+        handle_response_openai_compat(response).await
+    }
+}
+
+#[async_trait]
+impl Provider for LiteLLMProvider {
+    fn metadata() -> ProviderMetadata {
+        ProviderMetadata::new(
+            "litellm",
+            "LiteLLM",
+            "LiteLLM proxy supporting multiple models with automatic prompt caching",
+            LITELLM_DEFAULT_MODEL,
+            vec![],
+            LITELLM_DOC_URL,
+            vec![
+                ConfigKey::new("LITELLM_API_KEY", false, true, None),
+                ConfigKey::new("LITELLM_HOST", true, false, Some("http://localhost:4000")),
+                ConfigKey::new(
+                    "LITELLM_BASE_PATH",
+                    true,
+                    false,
+                    Some("v1/chat/completions"),
+                ),
+                ConfigKey::new("LITELLM_CUSTOM_HEADERS", false, true, None),
+                ConfigKey::new("LITELLM_TIMEOUT", false, false, Some("600")),
+            ],
+        )
+    }
+
+    fn get_model_config(&self) -> ModelConfig {
+        self.model.clone()
+    }
+
+    #[tracing::instrument(skip_all, name = "provider_complete")]
+    async fn complete(
+        &self,
+        system: &str,
+        messages: &[Message],
+        tools: &[Tool],
+    ) -> Result<(Message, ProviderUsage), ProviderError> {
+        let mut payload = super::formats::openai::create_request(
+            &self.model,
+            system,
+            messages,
+            tools,
+            &ImageFormat::OpenAi,
+        )?;
+
+        if self.supports_cache_control() {
+            payload = update_request_for_cache_control(&payload);
+        }
+
+        let response = self.post(payload.clone()).await?;
+
+        let message = super::formats::openai::response_to_message(response.clone())?;
+        let usage = match super::formats::openai::get_usage(&response) {
+            Ok(usage) => usage,
+            Err(ProviderError::UsageError(e)) => {
+                tracing::debug!("Failed to get usage data: {}", e);
+                Usage::default()
+            }
+            Err(e) => return Err(e),
+        };
+        let model = get_model(&response);
+        emit_debug_trace(&self.model, &payload, &response, &usage);
+        Ok((message, ProviderUsage::new(model, usage)))
+    }
+
+    fn supports_embeddings(&self) -> bool {
+        true
+    }
+
+    fn supports_cache_control(&self) -> bool {
+        if let Ok(models) = tokio::task::block_in_place(|| {
+            tokio::runtime::Handle::current().block_on(self.fetch_models())
```

Comment:
should we cache this? supports_cache_control sounds innocent enough I wouldn't expect to have it do a network call